# ABOUT IISD-ELA Chlorobia cyc2 project
Copyright Jackson M. Tsuji, Neufeld Research Group, 2019

This Zenodo repository describes the steps/code used to perform bioinformatics analysis for the paper by Tsuji and colleagues, 2019 (see [preprint here](_____)). Descriptor files in this repo are in the Markdown (`.md`) format and begin with `ABOUT`. Other files include scripts and raw data files. The information is organized as follows:

## `01_assembly_and_binning`
This folder describes the protocol used to produce assembled contigs and genome bins from the raw metagenomic datasets used for this paper. Two separate assembly/binning runs were performed, described in `ABOUT_lake_metagenomes.md` and `ABOUT_enrichment_metagenomes.md` for the lake samples and enrichment cultures, respectively.

## Folder `02_tables_and_figures`
This folder describes the protocol used to generate each table and figure in the main manuscript. ABOUT files exist for each.

## Folder `03_supplemental_tables_and_figures`
This folder describes the protocol used to generate each table, figure, or file in the supplementary materials not already described above. ABOUT files exist for each.

## Folder `04_extra_files`
This folder contains helpful supporting files outside those in the supplementary information, for users interested in the nitty gritty details of how the code was executed.

I hope this is a helpful resource for the scientific community, both in the critique of my work and in further development of bioinformatics. Enjoy!
Jackson M. Tsuji

